[PWGHF] Add KFParticle Quality Assurance in XicToXiPi candidate creator

Rrantu opened the pull request:
Evaluate the resolution of the particles reconstructed using the KFParticle method.

vkucera commented:
Hi @Rrantu , such a huge PR deserves some comments.
- First of all, what is your goal?
- Second, why do you need so many variables, new structs, histograms and tables?
- Why do you need to do further MC matching?
- How much more memory does the extra stuff cost?
- Last but not least, fix the issues first, only then make the PR ready.

Rrantu commented:
Hi @vkucera , I'm trying to check the resolution of the particles reconstructed using the KFParticle method. For this, I need the information saved in runXic0Omegac0Creator function, Therefore, I’ve added a new tree within the runXic0Omegac0Creator function to store these variables.
Because KFParticle reconstruction starts from V0s, my MC matching also needs to begin from V0s. So, I’ve written a new process function to read the new tree and do the MC matching accordingly.
I’m not exactly sure how much memory this will consume, but it should be minimal because the new tree is relatively small.
And for the O2 linter issues, they were not introduced by my changes.

zhangbiao-phy commented:
hi @Rrantu, Thanks for the implementation! I have a few comments. See the review

zhangbiao-phy commented on PWGHF/TableProducer/candidateCreatorXic0Omegac0.cxx:
```diff
@@ -2386,6 +2639,392 @@ struct HfCandidateCreatorXic0Omegac0Mc {
+
+    float lambdaXDelta;
+    float lambdaYDelta;
+    float lambdaZDelta;
+
+    float xiBachelorXDelta;
+    float xiBachelorYDelta;
+    float xiBachelorZDelta;
+    float xiBachelorPt;
+    float xiBachelorXPull;
+    float xiBachelorYPull;
+    float xiBachelorZPull;
+
+    float xiXDelta;
+    float xiYDelta;
+    float xiZDelta;
+    float xiPt;
+    float xiXPull;
+    float xiYPull;
+    float xiZPull;
+
+    float xic0BachelorXDelta;
+    float xic0BachelorYDelta;
+    float xic0BachelorZDelta;
+    float xic0BachelorPt;
+    float xic0BachelorXPull;
+    float xic0BachelorYPull;
+    float xic0BachelorZPull;
+
+    float xic0XDelta;
+    float xic0YDelta;
+    float xic0ZDelta;
+    float xic0Pt;
+    float xic0XPull;
+    float xic0YPull;
+    float xic0ZPull;
+
+    float cascXDelta;
+    float cascYDelta;
+    float cascZDelta;
```
Can you simplify the initialization and assignment of these variables? You can do it all inside the candidate loop. For example: `auto v0MassPull = (candidate.invMassLambda() - MassLambda0) / candidate.invMassV0Sig();`

Rrantu replied:
Thanks for the feedback, I’ve updated the code accordingly.

zhangbiao-phy commented:
> Hi @vkucera , I'm trying to check the resolution of the particles reconstructed using the KFParticle method. For this, I need the information saved in runXic0Omegac0Creator function, Therefore, I’ve added a new tree within the runXic0Omegac0Creator function to store these variables. Because KFParticle reconstruction starts from V0s, my MC matching also needs to begin from V0s. So, I’ve written a new process function to read the new tree and do the MC matching accordingly. I’m not exactly sure how much memory this will consume, but it should be minimal because the new tree is relatively small. And for the O2 linter issues, they were not introduced by my changes.

Hi @Rrantu, Thanks! Please fix the O2 linter issues next time!

fgrosa commented on PWGHF/TableProducer/candidateCreatorXic0Omegac0.cxx:
```diff
@@ -236,6 +237,65 @@ struct HfCandidateCreatorXic0Omegac0 {
     float chi2MassCasc;
     float etaXic;
   } kfXic0Candidate;
+
```
following @vkucera, why do you need a struct here? Can't you directly fill the table?

zhangbiao-phy replied:
I agree, it's useless, we can remove it and fill the table directly

Rrantu replied:
Thank you for your feedback, I've update the code as you suggested.

fgrosa commented on PWGHF/DataModel/CandidateReconstructionTables.h:
```diff
@@ -1368,6 +1368,58 @@ DECLARE_SOA_COLUMN(MassV0Chi2OverNdf, massV0Chi2OverNdf, float);
 DECLARE_SOA_COLUMN(CascRejectInvmass, cascRejectInvmass, float);
 
+// Kf QA results:
+DECLARE_SOA_COLUMN(InvMassV0Sig, invMassV0Sig, float);
+DECLARE_SOA_COLUMN(InvMassXiSig, invMassXiSig, float);
+DECLARE_SOA_COLUMN(InvMassXic0Sig, invMassXic0Sig, float);
```
What is the difference between these invariant masses and those in the usual table?

Rrantu replied:
Hi @fgrosa, Thanks for your comments, these columns contain the uncertainty of the invariant mass, which is not included in the default table.

fgrosa replied:
Thanks for the clarification! Maybe use a more meaningful name (to make explicit that it is an uncertainty)

Rrantu replied:
Thanks, I've changed the column name.

fgrosa commented on PWGHF/TableProducer/candidateCreatorXic0Omegac0.cxx:
```diff
@@ -1576,6 +1638,63 @@ struct HfCandidateCreatorXic0Omegac0 {
       registry.fill(HIST("hDcaXYCascadeToPVKf"), kfXic0Candidate.kfDcaXYCascToPv);
 
+      // QA
+      kfXic0Qa.v0DauPosX = kfPos.GetX();
```
why filling first the struct and then the table? Can't you directly fill the table?

fgrosa commented on PWGHF/TableProducer/candidateCreatorXic0Omegac0.cxx:
```diff
@@ -1896,6 +2027,128 @@ struct HfCandidateCreatorXic0Omegac0Mc {
+      registry.add("hXiMassPullVsPt", "m_{PULL}(#Xi^{-}) vs. p_{T}", HistType::kTH2F, {{20, 0., 20.}, {2000, -10., 10.}});
+      registry.add("hXic0MassPullVsPt", "m_{PULL}(#Xic0) vs. p_{T}", HistType::kTH2F, {{20, 0., 20.}, {2000, -10., 10.}});
+      // delta
+      registry.add("hV0DauPosXDelta", "x^{p} - x^{MC}", kTH1F, {{2000, -10., 10.}});
```
Replace all `TH1F` with `TH1D`
Do you also need all these histograms since you are filling a table with the same variables?

fgrosa commented on PWGHF/TableProducer/candidateCreatorXic0Omegac0.cxx:
```diff
@@ -1617,6 +1736,18 @@ struct HfCandidateCreatorXic0Omegac0 {
                          v0NDF, cascNDF, charmbaryonNDF, v0Ndfm, cascNdfm,
                          v0Chi2OverNdf, cascChi2OverNdf, charmbaryonChi2OverNdf, v0Chi2OverNdfm, cascChi2OverNdfm);
 
+      // fill QA table
```
This should be filled only if the QA is enabled and not by default

Rrantu replied:
Hi @fgrosa, I’ve made the changes as you suggested. Could you please approve them?

fgrosa reviewed (changes requested):
Hi @Rrantu, thanks for the development! I have however a few comments (some of them follow the doubts raised by @vkucera). You can always retrieve the info about the matching from the "default" table and, if you want to add more flags for matching for intermediate states, you can do it in the MC process function as well without duplicating

alibuild commented:
Error while checking build/O2Physics/o2 for 3268c45cc9ce86719a57bc8deecd887d675f8c40 at 2025-04-25 14:41:
```
## sw/BUILD/O2Physics-latest/log
/sw/SOURCES/O2Physics/10972-slc9_x86-64/0/PWGHF/TableProducer/candidateCreatorXic0Omegac0.cxx:2579:86: error: 'const struct o2::soa::Table<o2::aod::Hash<3036674320>, o2::aod::Hash<295670893>, o2::aod::Hash<2286545062> >::TableIteratorBase<o2::soa::DefaultIndexPolicy, o2::soa::Table<o2::aod::Hash<3036674320>, o2::aod::Hash<295670893>, o2::aod::Hash<2286545062> > >' has no member named 'invMassV0Sig'; did you mean 'invMassV0Err'?
/sw/SOURCES/O2Physics/10972-slc9_x86-64/0/PWGHF/TableProducer/candidateCreatorXic0Omegac0.cxx:2660:89: error: 'const struct o2::soa::Table<o2::aod::Hash<3036674320>, o2::aod::Hash<295670893>, o2::aod::Hash<2286545062> >::TableIteratorBase<o2::soa::DefaultIndexPolicy, o2::soa::Table<o2::aod::Hash<3036674320>, o2::aod::Hash<295670893>, o2::aod::Hash<2286545062> > >' has no member named 'invMassXiSig'; did you mean 'invMassXiErr'?
/sw/SOURCES/O2Physics/10972-slc9_x86-64/0/PWGHF/TableProducer/candidateCreatorXic0Omegac0.cxx:2727:94: error: 'const struct o2::soa::Table<o2::aod::Hash<3036674320>, o2::aod::Hash<295670893>, o2::aod::Hash<2286545062> >::TableIteratorBase<o2::soa::DefaultIndexPolicy, o2::soa::Table<o2::aod::Hash<3036674320>, o2::aod::Hash<295670893>, o2::aod::Hash<2286545062> > >' has no member named 'invMassXic0Sig'; did you mean 'invMassXic0Err'?
ninja: build stopped: subcommand failed.
```
Full log [here](https://ali-ci.cern.ch/alice-build-logs/AliceO2Group/O2Physics/10972/3268c45cc9ce86719a57bc8deecd887d675f8c40/build_O2Physics_o2/pretty.html).

vkucera commented on PWGHF/TableProducer/candidateCreatorXic0Omegac0.cxx:
```diff
@@ -1896,6 +1913,128 @@ struct HfCandidateCreatorXic0Omegac0Mc {
+    // QA
+    if (doprocessMcXicToXiPiKfQa) {
+      // mass over pt
+      registry.add("hV0MassPullVsPt", "m_{PULL}(V0) vs. p_{T}", HistType::kTH2D, {{20, 0., 20.}, {2000, -10., 10.}});
+      registry.add("hXiMassPullVsPt", "m_{PULL}(#Xi^{-}) vs. p_{T}", HistType::kTH2D, {{20, 0., 20.}, {2000, -10., 10.}});
+      registry.add("hXic0MassPullVsPt", "m_{PULL}(#Xic0) vs. p_{T}", HistType::kTH2D, {{20, 0., 20.}, {2000, -10., 10.}});
```
Avoid hard-coded duplication of axis definitions by using axis objects.

vkucera commented on PWGHF/TableProducer/candidateCreatorXic0Omegac0.cxx:
```diff
@@ -1896,6 +1913,128 @@ struct HfCandidateCreatorXic0Omegac0Mc {
+      registry.add("hXic0YPull", "y^{PULL}(#Xi_{c}^{0})", kTH1D, {{4000, -20., 20.}});
+      registry.add("hXic0ZPull", "z^{PULL}(#Xi_{c}^{0})", kTH1D, {{4000, -20., 20.}});
+      // pull over pt
+      registry.add("hV0DauPosXPullVsPt", "x_{PULL} vs. p_{T}", HistType::kTH2D, {{20, 0., 20.}, {4000, -20., 20.}});
+      registry.add("hV0DauPosYPullVsPt", "y_{PULL} vs. p_{T}", HistType::kTH2D, {{20, 0., 20.}, {4000, -20., 20.}});
+      registry.add("hV0DauPosZPullVsPt", "z_{PULL} vs. p_{T}", HistType::kTH2D, {{20, 0., 20.}, {4000, -20., 20.}});
```
That's a lot of bins. Please evaluate the extra RAM consumption when running.

Rrantu replied:
Thanks for the feedback. I have reduced the number of bins.

vkucera replied:
Let's see if it runs fine, but please check the extra memory consumption. It's still a lot of bins.

vkucera commented on PWGHF/TableProducer/candidateCreatorXic0Omegac0.cxx:
```diff
@@ -2386,6 +2525,261 @@ struct HfCandidateCreatorXic0Omegac0Mc {
   } // close process
 
+  template <o2::hf_centrality::CentralityEstimator centEstimator, int decayChannel, typename TMyRecoCand>
+  void runXic0Omegac0McQa(TMyRecoCand const& candidates,
+                          MyTracksWMc const&,
+                          aod::McParticles const& mcParticles,
+                          BCsInfo const&)
+  {
```
Can the body of this function be done inside `runXic0Omegac0Mc` instead?

vkucera replied:
No, they  don't:
```c++
  void runXic0Omegac0Mc(TMyRecoCand const& candidates,
                        MyTracksWMc const&,
                        aod::McParticles const& mcParticles,
                        Colls const& collsWithMcLabels,
                        McCollisions const& mcCollisions,
                        BCsInfo const&)
```

Rrantu replied:
Sorry for the earlier unclear explanation. I didn’t use runXic0Omegac0Mc because it fills MCGen and MCRec tables that I don’t need, and they consume a large amount of memory.
Also, the matching in runXic0Omegac0Mc goes from Xic0 to Λ0, while for my QA I actually need matching from Λ0 to Xic0, so the matching setup is different.
Since runXic0Omegac0Mc is used in many other analyses, modifying it directly could affect others. So I added an independent runXic0Omegac0McQA function to minimize the impact.

vkucera commented on PWGHF/TableProducer/candidateCreatorXic0Omegac0.cxx:
```diff
@@ -2386,6 +2525,261 @@ struct HfCandidateCreatorXic0Omegac0Mc {
     } // close loop on MCCollisions
   } // close process
 
+  template <o2::hf_centrality::CentralityEstimator centEstimator, int decayChannel, typename TMyRecoCand>
```
```suggestion
  template <o2::hf_centrality::CentralityEstimator centEstimator, aod::hf_cand_xic0_omegac0::DecayType decayChannel, typename TMyRecoCand>
```
Please fix this elsewhere too.

vkucera commented on PWGHF/TableProducer/candidateCreatorXic0Omegac0.cxx:
```diff
@@ -2386,6 +2525,261 @@ struct HfCandidateCreatorXic0Omegac0Mc {
+      auto mcXic0Bachelor = arrayDaughters[0].mcParticle();
+
+      // Xic0 -> xi pi matching
+      if constexpr (decayChannel != aod::hf_cand_xic0_omegac0::DecayType::XiczeroToXiPi) {
+        LOGF(info, "ERROR: Quality validation is restricted to Xic0 → Xi Pi decay processes at this stage");
+      }
```
I guess you want to throw a `LOGF(fatal,` here?

Rrantu replied:
Hi @vkucera, I’ve made the changes as you suggested. Could you please approve them?

vkucera commented on PWGHF/TableProducer/candidateCreatorXic0Omegac0.cxx:
```diff
@@ -2386,6 +2525,261 @@ struct HfCandidateCreatorXic0Omegac0Mc {
+        if (flag != 0) {
+          auto particle = mcParticles.rawIteratorAt(indexRecCharmBaryon);
+        }
+        if (debug == McMatchFlag::CascUnmatched || debug == McMatchFlag::V0Unmatched) {
+          LOGF(info, "WARNING: Xic0ToXiPi decays in the expected final state but the condition on the intermediate states are not fulfilled");
+        }
```
Use logging and severity properly. Do not write `WARNING`, use the correct logging level.
What do you want to do when his happens?

vkucera replied:
Also what is the purpose of the `debug` variable? Usually, it's a flag for logging level which is not the case here.

Rrantu replied:
Hi @vkucera , thanks for your feedback. I’ve updated the code accordingly, and the debug variable wasn’t needed, so I removed it.

vkucera replied:
It's still there.

Rrantu replied:
Sorry, Now the debug variable is properly removed.

vkucera commented on PWGHF/TableProducer/candidateCreatorXic0Omegac0.cxx:
```diff
@@ -118,6 +119,7 @@ struct HfCandidateCreatorXic0Omegac0 {
   Configurable<int> kfConstructMethod{"kfConstructMethod", 2, "KF Construct Method"};
   Configurable<bool> kfUseV0MassConstraint{"kfUseV0MassConstraint", false, "KF: use Lambda mass constraint"};
   Configurable<bool> kfUseCascadeMassConstraint{"kfUseCascadeMassConstraint", false, "KF: use Cascade mass constraint"};
+  Configurable<bool> kfResolutionQA{"kfResolutionQA", true, "KF: KFParticle Quality Assurance"};
```
```suggestion
  Configurable<bool> kfResolutionQA{"kfResolutionQA", false, "KF: KFParticle Quality Assurance"};
```
Do not make new features activated by default unless they are needed for a correct functionality of the code.

vkucera commented on PWGHF/TableProducer/candidateCreatorXic0Omegac0.cxx:
```diff
@@ -2386,6 +2536,239 @@ struct HfCandidateCreatorXic0Omegac0Mc {
+      auto mcXic0Bachelor = arrayDaughters[0].mcParticle();
+
+      // Xic0 -> xi pi matching
+      if constexpr (decayChannel != aod::hf_cand_xic0_omegac0::DecayType::XiczeroToXiPi) {
+        LOGF(fatal, "ERROR: Quality validation is restricted to Xic0 → Xi Pi decay processes at this stage");
+      }
```
```suggestion
      if constexpr (decayChannel != aod::hf_cand_xic0_omegac0::DecayType::XiczeroToXiPi) {
        LOGF(fatal, "Quality validation is restricted to Xic0 → Xi Pi decay processes at this stage");
      }
```
- The severity level is already defined with `fatal`. Adding a text "ERROR" is not only redundant but also misleading because `error` is a different severity level.
- Any sanity check should be done asap before any unrelated calculations.
- The result of this check is known at compile time. Making it fail at run time is too late. 
- What is the point of making `decayChannel` a parameter when only one value is valid?

Rrantu replied:
Thank you for the suggestion. The decayChannel selection was already applied before the match, so I have removed the check accordingly.

fgrosa commented:
@vkucera @Rrantu have all the comments been implemented? Can we merge this?

Rrantu commented:
> @vkucera @Rrantu have all the comments been implemented? Can we merge this?
I believe I’ve addressed all the comments. Please let me know if anything else is needed before merging.
